lib/client.js

```diff
@@ -20,58 +20,82 @@ module.exports = function (dependencies) {
   function Client (options) {
     this.config = config(options);
     this.healthCheckInterval = setInterval(() => {
-         if (this.session && !this.session.destroyed) {
-             this.session.ping((error, duration) => {
-                 if (error) {
-                     logger("No Ping response after " + duration + " ms");
-                 }
-                 logger("Ping response after " + duration + " ms");
-             })
-         }
-     }, this.config.heartBeat).unref();
+      if (this.session && !this.session.closed && !this.session.destroyed && !this.isDestroyed) {
+        this.session.ping((error, duration) => {
+          if (error) {
+            logger("No Ping response after " + duration + " ms");
+          }
+          logger("Ping response after " + duration + " ms");
+        });
+      }
+    }, this.config.heartBeat).unref();
   }
 
   Client.prototype.write = function write (notification, device, count) {
+    if (this.isDestroyed) {
+      return Promise.resolve({ device, error: new VError("client is destroyed") });
+    }
+
     // Connect session
-    if (!this.session || this.session.destroyed) {
-      this.session = http2.connect(`https://${this.config.address}`, this.config);
+    if (!this.session || this.session.closed || this.session.destroyed) {
+      const session = this.session = http2.connect(`https://${this.config.address}`, this.config);
+
+      const destroy = (callback) => {
+        if (this.session === session) {
+          this.session = null;
+        }
+        if (!session.destroyed) {
+          session.destroy();
+        }
+        if (callback) {
```

Comment:
I think that destroying a session and the session.destroyed flag being set may not be guaranteed to be synchronous

Reply:
This verification is only to avoid calling destroy function twice. According to the documentation, the session is immediately destroyed and since there is no callback for this function I believe it is synchronous.
https://nodejs.org/api/http2.html#http2_http2session_destroy_error_code